server/src/main/java/org/elasticsearch/plugins/NodeTerminationHandlerPlugin.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.client.internal.node.NodeClient;
+import org.elasticsearch.env.NodeEnvironment;
+
+/**
+ * A plugin that performs operations when the node is shut down, i.e. in response to a sigterm.
+ */
+public interface NodeTerminationHandlerPlugin {
+
+    void handleTerminationSignal(NodeClient client, NodeEnvironment nodeEnvironment);
```

Comment:
Why are the client and env needed? Can you add a javadoc describing their purpose?

Reply:
As shown in [the example implementation](https://gist.github.com/gwbrown/4d4c28f40f031563f35355b0bba68497), a `Client` is needed so that we can issue requests, and `NodeEnvironment` provides the node ID of the current node. We could trim that down to just the node ID depending on how hard we want to tailor this to our particular use case.